---
title: ""
output:
  html_document:
    theme: united
    df_print: kable
  pdf_document: default
---

```{r fxns, include = FALSE}
library('htmlTable')
make_list <- function(items) {
  paste0("<ul>", sprintf('<li>%s</li>', items), '</ul>', collapse = '')
}
make_href <- function(url, text){
  paste0("<div style='text-align:center; font-size:large; font-weight:bold'><a href=", url ,">", text, "</a></div>")
}
make_href2 <- function(url, text){
  paste0("<a href=", url ,">", text, "</a>")
}
process_entry <- function(dat) {
  if (grepl(pattern = "https://satijalab.org/img/vignette_images", x = dat$image)) {
    img <- paste0('![](', dat$image, '){width=3000px}')
  } else if (grepl(pattern = "assets/", x= dat$image)) {
    img <- paste0('![](', dat$image, '){width=3000px}') 
  } else {
    img <- paste0('![](', '../output/images/', dat$image, '){width=3000px}')
  }
  if (grepl(pattern = "https://satijalab.org/", x = dat$name)) {
    link <- dat$name
  } else {
    link <- paste0(dat$name, ".html")
  }
  go.button <- paste0('<a class="btn btn-primary" href="', link, '" role="button" style="width:100%">GO</a>')
  data.frame(
    title = make_href(url = link, text = dat$title),
    img = img,
    desc = dat$summary,
    btn = go.button
  )
}
process_wrapper_entry <- function(dat) {
  data.frame(
    Package = dat$title,
    Vignette = make_href2(url = dat$link, text = dat$title),
    Reference = make_href2(url = dat$reference, text = dat$citation),
    Source = make_href2(url = dat$source, text = dat$source)
  )
}
make_vignette_card_section <- function(vdat, cat) {
  vignettes <- vdat[[cat]]$vignettes
  dat <- data.frame(title = character(), img = character(), desc = character())
  for (v in 1:length(x = vignettes)) {
    dat <- rbind(dat, process_entry(vignettes[[v]]))
    if(nrow(x = dat) == 3 | v == length(x = vignettes)){
      colnames(dat) <- NULL
      dat <- t(dat)
      if (ncol(x = dat) == 2) {
        print(htmlTable(
          dat, 
          align = '|l|l|', 
          css.cell = "padding-left: .75em; width: 50%", 
          css.class = "two-column-htmltable"
        ))
      } else if (ncol(x = dat) == 1){
        print(htmlTable(
          dat, 
          align = '|l|', 
          css.cell = "padding-left: .75em; width: 100%", 
          css.class = "one-column-htmltable"
        ))
      } else {
        print(htmlTable(
          dat, 
          align = '|l|l|l|', 
          css.cell = "padding-left: .75em; width: 30%"
        ))
      }
      dat <- data.frame(title = character(), img = character(), desc = character())
    }
  }
}
```

```{r yaml, include = FALSE}
library(yaml)
vdat <- read_yaml(file = "vignettes.yaml")
```

```{=html}
<style>
.two-column-htmltable {
  width: 66.666%;
}
.one-column-htmltable {
  width: 33.333%;
}

</style>
```

# Introductory Vignettes

For new users of Seurat, we suggest starting with a guided walk through of a dataset of 2,700 Peripheral Blood Mononuclear Cells (PBMCs) made publicly available by 10X Genomics. This tutorial implements the major components of a standard unsupervised clustering workflow including QC and data filtration, calculation of high-variance genes, dimensional reduction, graph-based clustering, and the identification of cluster markers.

If you are looking to analyze multimodal data, we provide an analysis example to familiarize users with the Seurat infrastructure necessary for working with multiple modalities. This example features CITE-seq (RNA + protein) data but the general principles are extendable to arbitrary modalities. Similarly, for spatially-resolved data we work through a few examples of common analysis tasks such as interactive visualization and identifying spatially variable patterns of gene expression. Check out some of the possibilities in the vignettes below.

```{r results='asis', echo=FALSE, warning=FALSE, message = FALSE}
make_vignette_card_section(vdat = vdat, cat = 1)
```

# Data Integration

A core component of Seurat includes [computational methods](https://www.cell.com/cell/fulltext/S0092-8674(19)30559-8) for integrated analysis of single-cell datasets generated across different conditions, technologies, or species. As an example, we provide a guided walk through for integrating and comparing PBMC datasets generated under different stimulation conditions that should help familiarize new users with the general RNA-RNA integration workflow. For those looking for solutions to speed up these analysis or work with large datasets, we recommend exploring the integration with RPCA [vignette](integration_rpca.html) and the tips for integrating large datasets [vignette](integration_large_datasets.html). 

Seurat also provides related methods for mapping new query data to existing references. These can be useful for users looking to quickly analyze their data in the context of larger pre-established references without needing to integrate or 'correct' their underlying expression data. An an introduction to this workflow, we provide a example of generating and then mapping to an RNA reference in this [vignette](integration_mapping.html). Additionally, we provide [methods](https://satijalab.org/v4preprint) for utilizing multimodal references to improve cell type annotation and impute unmeasured query modalities. 


```{r results='asis', echo=FALSE, warning=FALSE, message = FALSE}
make_vignette_card_section(vdat = vdat, cat = 2)
```

# New Statistical Approaches

Seurat also offers novel statistical methods for analyzing single-cell data. These include [Weighted Nearest Neighbor Analysis (WNN)](https://satijalab.org/v4preprint), an approach for jointly analyzing multimodal single-cell data by learning a per-cell weighting of each data type. For improving the signal-to-noise ratio in single-cell perturbation screens, we provide [mixscape](https://www.biorxiv.org/content/10.1101/2020.06.28.175596v1), a method to help remove confounding sources of variation in these kinds of experiments. Finally, for an improved modeling approach to normalization and variance stabilization, we provide an interface to [sctransform](https://genomebiology.biomedcentral.com/articles/10.1186/s13059-019-1874-1).

```{r results='asis', echo=FALSE, warning=FALSE, message = FALSE}
make_vignette_card_section(vdat = vdat, cat = 3)
```

# Other 

Here we provide a series of short vignettes to demonstrate a number of features that are commonly used in Seurat. Weâ€™ve focused the vignettes around questions that we frequently receive from users. Click on a vignette to get started.

```{r results='asis', echo=FALSE, warning=FALSE, message = FALSE}
make_vignette_card_section(vdat = vdat, cat = 4)
```

# SeuratWrappers

In order to facilitate the use of community tools with Seurat, we provide the Seurat Wrappers package, which contains code to run other analysis tools on Seurat objects. For the initial release, we provide wrappers for a few packages in the table below but would encourage other package developers interested in interfacing with Seurat to check out our contributor guide [here](https://github.com/satijalab/seurat.wrappers/wiki/Submission-Process).

```{r results='asis', echo=FALSE, warning=FALSE, message = FALSE}
library(knitr)
library(kableExtra)
cat <- 5
vignettes <- vdat[[cat]]$vignettes
dat <- data.frame(Package = character(),  Vignette = character(), Reference = character(), Source = character())
for (v in 1:length(x = vignettes)) {
  dat <- rbind(dat, process_wrapper_entry(vignettes[[v]]))
}
dat %>%
  kable(format = "html", escape = FALSE) %>%
  kable_styling(bootstrap_options = c("striped", "hover"))
```
